src/core/painting/qgspainting.cpp

```diff
@@ -219,3 +222,36 @@ bool QgsPainting::drawTriangleUsingTexture( QPainter *painter, const QPolygonF &
 
   return true;
 }
+
+int QgsPainting::qtDefaultDpiX()
+{
+  return qt_defaultDpiX();
+}
+
+int QgsPainting::qtDefaultDpiY()
+{
+  return qt_defaultDpiY();
+}
+
+void QgsPainting::applyScaleFixForQPictureDpi( QPainter *painter )
+{
+  // QPicture makes an assumption that we drawing to it with system DPI.
+  // Then when being drawn, it scales the painter. The following call
+  // negates the effect. There is no way of setting QPicture's DPI.
+  // See QTBUG-20361
+  painter->scale( static_cast< double >( QgsPainting::qtDefaultDpiX() ) / painter->device()->logicalDpiX(),
+                  static_cast< double >( QgsPainting::qtDefaultDpiY() ) / painter->device()->logicalDpiY() );
+}
+
+void QgsPainting::drawPicture( QPainter *painter, const QPointF &point, const QPicture &picture )
+{
+  // QPicture makes an assumption that we drawing to it with system DPI.
+  // Then when being drawn, it scales the painter. The following call
+  // negates the effect. There is no way of setting QPicture's DPI.
+  // See QTBUG-20361
+  const double xScale = static_cast< double >( QgsPainting::qtDefaultDpiX() ) / painter->device()->logicalDpiX();
```

Comment:
Why not calling applyScaleFixForQPictureDpi and using save/restore instead ?

Reply:
There's considerable overhead in save/restore calls, and it's trivial to avoid them here.

Reply:
OK. 

Just out of curiosity, do you have any discussion/information about this?

Reply:
Just my own experiences profiling QGIS rendering. And it makes sense -- saving / restoring QPainter state requires storing a lot of properties, especially when clip masks are in place. In comparison scaling is an almost zero-cost operation.